common/testing/testlogger/testlogger.go

```diff
@@ -0,0 +1,504 @@
+	}
+	for _, opt := range opts {
+		opt(tl)
+	}
+	if tl.wrapped == nil {
+		ztl := zaptest.NewLogger(t,
+			zaptest.Level(tl.state.level),
+			zaptest.WrapOptions(
+				zap.AddStacktrace(zap.ErrorLevel),
+				zap.WithCaller(tl.state.logCaller),
+			),
+		)
+		tl.wrapped = log.NewZapLogger(ztl).Skip(1)
+	}
+
+	// Only possible with a *testing.T until *rapid.T supports `Cleanup`
+	if ct, ok := t.(CleanupCapableT); ok {
+		// NOTE(tim): We don't care about anything logged after the test completes. Sure, this is racy,
+		// but it reduces the likelihood that we see stupid errors due to testing.T.Logf race conditions...
+		ct.Cleanup(tl.Close)
+	}
+
+	return tl
+}
+
+// ResetUnexpectedErrors resets the unexpected error state, returning the previous value.
+func (tl *TestLogger) ResetUnexpectedErrors() bool {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	prevUnexpectedLogs := tl.state.mu.unexpectedErrs
+	tl.state.mu.unexpectedErrs = false
+	return prevUnexpectedLogs
+}
+
+// Expect instructs the logger to expect certain errors, as specified by the msg and tag arguments.
+// This is useful for ignoring certain errors that are expected.
+func (tl *TestLogger) Expect(level Level, msg string, tags ...tag.Tag) *Expectation {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	if tl.state.logExpectations {
+		tl.wrapped.Info(fmt.Sprintf("(%p) TestLogger::Expecting: '%s'\n", tl, msg))
+	}
+	return tl.addExpectationLocked(false, level, msg, tags...)
+}
+
+// DontExpect instructs the logger to _not_ expect certain errors, as specified by the msg and tag arguments.
+// This is useful for ensuring that certain errors _do not_ occur during a test.
+func (tl *TestLogger) DontExpect(level Level, msg string, tags ...tag.Tag) *Expectation {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	if tl.state.logExpectations {
+		tl.wrapped.Info(fmt.Sprintf("(%p) TestLogger::NotExpecting: '%s'\n", tl, msg))
+	}
+	return tl.addExpectationLocked(true, level, msg, tags...)
+}
+
+func (tl *TestLogger) addExpectationLocked(unexpected bool, level Level, msg string, tags ...tag.Tag) *Expectation {
+	e := &Expectation{
+		testLogger: tl,
+		lvl:        level,
+		unexpected: unexpected,
+	}
+	m := newMatcher(msg, tags, e)
+	e.e = tl.state.mu.expectations[level].PushBack(m)
+	return e
+}
+
+// Forget removes a previously registered expectation.
+// It will no longer be used to determine if a log message is expected or unexpected.
+func (tl *TestLogger) Forget(e *Expectation) {
+	tl.state.mu.Lock()
+	defer tl.state.mu.Unlock()
+	tl.state.mu.expectations[e.lvl].Remove(e.e)
+}
+
+func (tl *TestLogger) isUnexpected(level Level, msg string, tags []tag.Tag) bool {
+	expectations := tl.state.mu.expectations[level]
+	for e := expectations.Front(); e != nil; e = e.Next() {
+		m, ok := e.Value.(matcher)
+		if !ok {
+			tl.state.t.Fatalf("Bug in TestLogger: invalid %T value in matcher list", e.Value)
+		}
+		if m.Matches(msg, tags) {
+			return m.expectation.unexpected
+		}
```

Comment:
This had an `if len(pat.tags) > 0 {` check but AFAICT the behavior is the same if the tags are passed in, empty or not.